orderer/consensus/follower/follower_chain.go

```diff
@@ -51,16 +143,128 @@ func (*Chain) Errored() <-chan struct{} {
 }
 
 func (c *Chain) Start() {
-	//TODO skeleton
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.started || c.stopped {
+		c.logger.Debugf("Not starting because: started=%v, stopped=%v", c.started, c.stopped)
+		return
+	}
+
+	c.started = true
+
+	go c.run()
+
+	c.logger.Info("Started")
 }
 
 func (c *Chain) Halt() {
-	//TODO skeleton
+	c.halt()
+
+	select {
+	case <-c.doneChan:
+	}
+}
+
+func (c *Chain) halt() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.stopped {
+		c.logger.Debug("Already stopped")
+		return
+	}
+	c.stopped = true
+	close(c.stopChan)
+	c.logger.Info("Stopped")
 }
 
-// StatusReport returns the ClusterRelation & Status
+// StatusReport returns the ClusterRelation & Status.
 func (c *Chain) StatusReport() (types.ClusterRelation, types.Status) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	clusterRelation := types.ClusterRelationFollower
+	if c.joinBlock != nil {
+		if err := c.amIInChannel(c.joinBlock); err == nil {
+			clusterRelation = types.ClusterRelationMember
+		}
+	}
+
 	status := types.StatusActive
-	//TODO if (height is >= join-block.height) return "active"; else return "onboarding"
-	return types.ClusterRelationFollower, status
+	if (c.joinBlock != nil) && (c.joinBlock.Header.Number >= c.support.Height()) {
+		status = types.StatusOnBoarding
+	}
+
+	return clusterRelation, status
+}
+
+func (c *Chain) run() {
+	c.logger.Debug("The follower.Chain puller goroutine is starting")
+	pullRetryTicker := time.NewTicker(500 * time.Millisecond) // TODO move constant to options
+
+	err := c.pull()
+
+runLoop:
+	for err != nil {
+		c.logger.Debugf("Pull failed, going to try again in %v; error: %s", pullRetryTicker, err)
+
+		select {
+		case <-c.stopChan:
+			c.logger.Debug("Received a stop signal")
+			break runLoop
+
+		case <-pullRetryTicker.C:
+			err = c.pull()
+		}
+	}
+
+	pullRetryTicker.Stop()
+	close(c.doneChan)
+	c.logger.Debug("The follower.Chain puller goroutine is exiting")
+}
+
+// pull blocks from other orderers, until the join block, or until a config block the indicates the orderer has become
+// a member pf the cluster. When the follower.Chain's job is done, this method halts, triggers the creation of a new
+// consensus.Chain, and returns nil.
+func (c *Chain) pull() error {
+	if c.joinBlock != nil {
+		err := c.pullUpToJoin()
```

Comment:
consider to just have a function pointer as a field and just invoke the function pointer.
The function pointer can be set at initialization time.

Reply:
I really dislike function pointers for flow control. I think it makes the code harder to read and maintain. I think it is justified for decoupling components, like in the "CreateChain", and even there, I would have preferred an interface.